cupy/_core/dlpack.pyx

```diff
@@ -163,53 +233,79 @@ cdef class DLPackMemory(memory.BaseMemory):
-        cdef DLManagedTensor* dlm_tensor
-
-        # sanity checks
-        if not cpython.PyCapsule_IsValid(dltensor, 'dltensor'):
-            raise ValueError('A DLPack tensor object cannot be consumed '
-                             'multiple times')
-        dlm_tensor = <DLManagedTensor*>cpython.PyCapsule_GetPointer(
-            dltensor, 'dltensor')
+        # First, take ownership of the contained DLManagedTensor(Versioned)
+        # by copying it over and setting the capsule name to "used".
+        if cpython.PyCapsule_IsValid(dltensor, CAPSULE_NAME):
+            # Take ownership of the memory:
+            self.dlm_tensor_unversioned = <DLManagedTensor*>(
+                cpython.PyCapsule_GetPointer(dltensor, CAPSULE_NAME))
+            cpython.PyCapsule_SetName(dltensor, USED_CAPSULE_NAME)
+
+            self.dl_tensor = &self.dlm_tensor_unversioned.dl_tensor
+        elif cpython.PyCapsule_IsValid(dltensor, CAPSULE_NAME_VER):
+            # Take ownership of the memory:
+            self.dlm_tensor_ver = <DLManagedTensorVersioned*>(
+                cpython.PyCapsule_GetPointer(dltensor, CAPSULE_NAME_VER))
+            cpython.PyCapsule_SetName(dltensor, USED_CAPSULE_NAME_VER)
+
+            # When we have a versioned tensor, we need to verify the version
+            # before further use.
+            if self.dlm_tensor_ver.version.major > 1:
+                raise BufferError("DLPack exported too new major version.")
+
+            # TODO(seberg): In principle we should raise an error if we got a
+            #     readonly buffer.  But that might be disruptive :(.
+            # if self.dlm_tensor_ver.flags & DLPACK_FLAG_BITMASK_READ_ONLY:
+            #  raise BufferError("Buffer is readonly, but...")
+
+            self.dl_tensor = &self.dlm_tensor_ver.dl_tensor
+        else:
+            # Be helpful in case a capsule is used twice somehow:
+            raise ValueError(
+                'A DLPack tensor object cannot be consumed multiple times '
+                f'(or object was not a DLPack capsule). Got: {dltensor!r}')
+
+        # Check if the device is compatible with the cupy runtime.
         if runtime._is_hip_environment:
-            if dlm_tensor.dl_tensor.device.device_type != kDLROCM:
+            if self.dl_tensor.device.device_type != kDLROCM:
                 raise RuntimeError('CuPy is built against ROCm/HIP, different '
                                    'from the backend that backs the incoming '
                                    'DLPack tensor')
         else:
-            if dlm_tensor.dl_tensor.device.device_type not in (
+            if self.dl_tensor.device.device_type not in (
                     kDLCUDA, kDLCUDAManaged):
                 raise RuntimeError('CuPy is built against CUDA, different '
                                    'from the backend that backs the incoming '
                                    'DLPack tensor')
 
-        self.dltensor = dltensor
```

Comment:
Q: This always confuses me. We don't hold onto the (used) capsule because if it's used its deleter is not kicked in, until self (`DLPackMemory`) has a 0 refcount?

Reply:
We transfer `dlm_tensor` ownership from the capsule to `DLPackMemory` (via renaming).  After that, the capsule serves no purpose anymore and can/should be deleted.

Reply:
I tried to make the pattern slightly more obvious, by adding a slightly weird:
```
try:
    create capsule
except:
    cleanup
    raise
else:
    # finalize initialization code that must not fail.
```
Where the else is mostly nice because it is a nice place to put the incref (the decref had a bug!).